### ISSUE TITLE
missing copy and move constructors and assignment operators

### ISSUE THREAD

**Issue by arturoc** [openframeworks/openFrameworks#4557]
There's several classes in the core and probably in the addons that should have copy and move constructors and assignment operators.

by the rule of 3 (the rule of 5 since c++11) everything that has a destructor needs mostly sure a copy and move constructor and assignment operator.

in most objects where we have destructors is just an empty virtual one which is usually fine but there's several objects that use a destructor to unassing themselves from an event. that means that at some point they registered to an event.

if those objects are copied with the default copy constructor the copies won't be registered to the event anymore. 

the best way to solve this rather than implementing those constructors which would be really verbose and error prone would be avoiding the destructor altogether.

a possible solution for the cases where the destructor is needed to unregister from an even would be to have some kind of polling event. an object that would register itself to an event and accumulate the events on a queue, then it could be queried to  check for new events, it would auto unregister it self from the event on destruction and do the right thing on copy or move.

the object containing it would be able to ask for new events on update or a similar call with an api similar to ofThreadChannel (it could even be just a new constructor of method for ofThreadChannel):

```
class MyClass{
public:
MyClass(){
    mousePolling.register(ofEvents().mouseMoved);
}
void update(){
    ofMouseEventArgs mouse;
    if(mousePolling.tryReceive(mouse)){
        // do something with mouse movement
    }
}
}
```

this particular example probably doesn't make much sense cause there's already functions to check the state of the mouse but it will be useful for other cases.

**Comment by ofTheo** [comment]
think getting rid of destructors would be good idea. 
It does look like with the current event system there is no way to know if an object is still around to receive events.  

Either way moving to an event system which allows for automatic unregistration would be super helpful. 
I think so many bugs come from forgetting to unregister. 

If we are going to change the api, we should definitely think of ways to make it a bit friendlier. 
I sort of like the option to both register a callback for the event and to poll it as well ( not sure if this is possible ) 

If we had either a base class ( ofBaseReceivesEvents ) or an object ofEventListener we could do automatic unregistration of the event. 

ie:

```
    //in ofApp.h 
    ofEventListener listener; 
 
    //in ofApp.cpp 
    listener.add(GameEvent::events, this, &ofApp::gameEvent);
```

then on destruction of ofApp (or whichever class) the event listener object notifies the event to stop firing to it. 

Could something like this work?

**Comment by arturoc** [comment]
yes i've tried that too but and it avoids the unregistration but you still have the problem with copy constructors. if you do:

```
class GameObject{
    ofEventListener listener; 
public:
    GameObject(){
        listener.add(GameEvent::events, this, &ofApp::gameEvent);
    }
```
and then:

```
GameObject originalObj;
GameObject obj(originalObj);
```

obj won't be registered to the event

**Comment by arturoc** [comment]
same thing if you try to copy/move the object into a vector

**Comment by ofTheo** [comment]
oh yeah :( 
was thinking we could use a smart pointer inside ofEventListener - but its the raw pointer "this" part which messes us up. 

hmmm - its a hard problem :)

**Comment by ofTheo** [comment]
Anyway to get around it with a base class? 
Pass something different rather than this? 

```
class GameObject : public ofBaseRecievesEvents{
public:
    GameObject(){
        addListener(GameEvent::events, ofBaseRecievesEvents::getListenerAddress(), & GameObject::gameEvent);
    }
```

clutching at straws here :)

**Comment by Spacechild1** [comment]
Hi, I'm generally missing move constructors and assignments in some classes. Passing temporary objects as arguments to class methods will often result in unnecessary copying (like in myPixels = myClass.getPixels). ofImage has at least a move constructor, but the method setFromPixels for example has no overload for rvalue References. Is this going to change in the near future?

**Comment by arturoc** [comment]
yes, this issue is specifically about classes that need them because we have registered events but for other classes (mostly data containers as you describe) would be great to have them too where it makes sense (trying to avoid them if the class default constructors already does the right thing) so if you want to send a PR it would be really useful
